### PR TITLE
[WasmFS] Add [f]truncate syscalls

### PR DESCRIPTION
**Pull request by kripken**
Add `setSize()` to `DataFile`.

Fix the JS API helper `_wasmfs_write_file` to create files in read-write
mode and not write-only. This is just for tests, and seems like a better
default there, and is necessary for this test to pass.

### CODE REVIEW

File: system/lib/wasmfs/syscalls.cpp [hunk 1]
```diff
@@ -1048,4 +1048,60 @@ long __syscall_faccessat(long dirfd, long path, long amode, long flags) {
 
   return 0;
 }
+
+static off_t combineOffParts(long low, long high) {
+  return (off_t(high) << 32) | off_t(low);
+}
+
+long doTruncate(std::shared_ptr<File>& file, long low, long high) {
```
Comment (sbc100) [inline system/lib/wasmfs/syscalls.cpp:1056]: static

**Review by sbc100 — APPROVED** [review]

File: system/lib/wasmfs/syscalls.cpp [hunk 2]
```diff
@@ -1048,4 +1048,60 @@ long __syscall_faccessat(long dirfd, long path, long amode, long flags) {
...
+
+long __syscall_ftruncate64(long fd, long low, long high) {
+  auto openFile = wasmFS.getFileTable().locked().getEntry(fd);
+  if (!openFile) {
+    return -EBADF;
+  }
+  auto ret = doTruncate(openFile->locked().getFile(), low, high);
+  // XXX It is not clear from the docs why ftruncate would differ from
+  //     truncate here. However, on Linux this definitely happens, and the old
+  //     FS matches that as well, so do the same here.
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:1101]: Not only is it unclear, it seems to directly contradict the docs. It is really worth being bug-compatible with linux in this case?

**Review by tlively — COMMENTED** [review]
LGTM % comments

File: system/lib/wasmfs/backends/proxied_file_backend.cpp [hunk 3]
```diff
@@ -52,6 +52,11 @@ class ProxiedFile : public DataFile {
     return result;
   }
 
+  void setSize(size_t size) override {
+    // TODO
+    abort();
```
[the remark this answers is not in the document]
Comment (kripken) [reply system/lib/wasmfs/backends/proxied_file_backend.cpp:57]: Ah, I just commented in another PR, I think an abort is better as assertions don't happen in release builds. But yeah, not having a message is less good... We could add a macro perhaps that does both, I pushed one now, PTAL.

File: system/lib/wasmfs/syscalls.cpp (hunk 2 above)
Comment (kripken) [reply system/lib/wasmfs/syscalls.cpp:1101]: Good question. I guess I'm not 100% sure it's a bug and not something we're missing in the docs. But if we are both certain it is a linux bug, perhaps it's fine to ignore?

File: system/lib/wasmfs/support.cpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#include <iostream>
+
+#include "support.h"
+
+namespace wasmfs {
+
+void handle_unreachable(const char* msg, const char* file, unsigned line) {
+#ifndef NDEBUG
+  if (msg) {
+    std::cerr << msg << "\n";
```
Comment (tlively) [inline system/lib/wasmfs/support.cpp:15]: Should we use `emscripten_console_errorf` here to avoid using any additional file system code?

**Review by tlively — APPROVED** [review]

File: system/lib/wasmfs/syscalls.cpp (hunk 2 above)
Comment (tlively) [reply system/lib/wasmfs/syscalls.cpp:1101]: I couldn't find any note of this on the internet. I guess the safer thing to do is match linux and the existing system, and at least this is a great comment explaining why.

File: system/lib/wasmfs/backends/proxied_file_backend.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (tlively) [reply system/lib/wasmfs/backends/proxied_file_backend.cpp]: Nice!

File: system/lib/wasmfs/support.cpp (hunk 4 above)
Comment (kripken) [reply system/lib/wasmfs/support.cpp:29]: Very good point! Changed to that.

**Pull request merged**